Calibration/TkAlCaRecoProducers/plugins/CalibrationTrackSelectorFromDetIdList.cc

```diff
@@ -108,22 +107,20 @@ void CalibrationTrackSelectorFromDetIdList::produce(edm::Event& iEvent, const ed
       }
 
       // here there will be the selection
-      hits.push_back(hit->clone());
+      hits.emplace_back(hit->clone());
```

Comment:
hi @Sam-Harper @mmusich - this copy is not needed as the hits are again copied later in the code. perhaps just a std::vector<const TrackingRecHit*> filled with 'hit' rather than 'hit->clone()' would be a more efficient fix.

Reply:
hmm I'm not sure I follow. I dont see where it copies. I can see places where it looks like it copies but doesnt. [ownHits](https://github.com/cms-sw/cmssw/blob/dd4bef92af985d344e135b3ff585d77f1ba5cc3d/Calibration/TkAlCaRecoProducers/plugins/CalibrationTrackSelectorFromDetIdList.cc#L139) is a [edm::OwnVector](https://github.com/cms-sw/cmssw/blob/master/DataFormats/TrackCandidate/interface/TrackCandidate.h#L26) which we pass pointers into (and their ownership) not a vector.  And when we create the [TrackCandidate](https://github.com/cms-sw/cmssw/blob/dd4bef92af985d344e135b3ff585d77f1ba5cc3d/Calibration/TkAlCaRecoProducers/plugins/CalibrationTrackSelectorFromDetIdList.cc#L145) its also not a copy, it moves the own vector into it (well swaps it).  Is it due to the return of that function, I though return value optimization would take care of that although I'm far from an expert in these things.

Regardless which of us is correct,  you raise a good point, we should have a vector of pointers here and only if we actually save a track should we start making copies, as its coded now its inefficient. I think I'll leave that as an exercise for @mmusich, my involvement is solely a minimal fix for the problems it was causing EGamma (well for CMS but Egamma was the canary)

Reply:
I wonder who really is aware of the four(!)  different semantics of OwnVector::push_back
(I was not).
https://cmssdt.cern.ch/dxr/CMSSW/source/DataFormats/Common/interface/OwnVector.h#288
In my opinion they are just confusing and obfuscating...
OwnVector is a remnant from the time when OO was supposed to "hide complexity" behind a common "simple", "natural" interface: failed! :-(look to its iterator)
anyhow this is how history goes.

Reply:
I followed up at #24759